src/acquisition/covid_hosp/common/database.py

```diff
@@ -172,6 +172,33 @@ def insert_dataset(self, publication_date, dataframe):
                        tuple(values) +
                        tuple(i[0] for i in self.additional_fields))
 
+    if hasattr(self, 'AGGREGATE_KEY_COLS'):
+      ak_cols = self.AGGREGATE_KEY_COLS
+
+      # restrict data to just the key columns and remove duplicate rows
+      ak_data = dataframe[ak_cols].drop_duplicates()
+      # cast types (bools and NULLables)
+      for col in ak_cols:
+        if col.startswith('is_'): # TODO: this is hacky af
+          ak_data[col] = (ak_data[col] == 'true')
+      ak_data = ak_data.to_numpy(na_value=None).tolist()
```

Comment:
maybe less hacky -- we already know the python dtype of each column, including the aggregate key columns
```suggestion
      dataframe_typemap = {
        name: dtype
        for name, _, dtype in dataframe_columns_and_types
      }
      # cast types (bools and NULLables)
      for col in ak_cols:
        ak_data[col] = dataframe_typemap[col](ak_data[col])
```

Reply:
lol, maybe i should have left my "hacky" comment out of this...  i did it that way because i didnt want to have to resort to using the `dataframe_columns_and_types` when the boolean column was the only one that needed massaging.

your suggestion is pretty good, but it needs an `apply()` and will probably have a few other hoops to jump through for NaN handling.  ill do that and add another commit.

Reply:
less hacky alternative has been pushed